annotation/tests/test_tasks_services.py

```diff
@@ -580,8 +789,1416 @@ def test_filter_tasks_db_no_files_or_jobs(
+
+
+@pytest.mark.parametrize(
+    (
+        "common_objs_ids",
+        "all_tasks_objs",
+        "id_mapping",
+        "expected_links",
+        "expected_children",
+    ),
+    (
+        (
+            (1, 2),
+            {
+                1: {"links": [{"to": 2}], "children": [3]},
+                2: {"links": [{"to": 3}], "children": []},
+            },
+            {(2,): 200, (3,): 300},
+            [[{"to": 2}], [{"to": 3}]],
+            [[3], []],
+        ),
+        (
+            (1,),
+            {1: {"links": [{"to": 2}], "children": []}},
+            {(2,): 200},
+            [[{"to": 2}]],
+            [[]],
+        ),
+        ((), {}, {}, [], []),
+    ),
+)
+def test_get_links_and_children(
+    common_objs_ids: Tuple[int],
+    all_tasks_objs: Dict[int, Dict[str, List[Dict[str, int]]]],
+    id_mapping: Dict[Tuple[int], int],
+    expected_links: List[List[Dict[str, int]]],
+    expected_children: List[List[int]],
+):
+    with patch("annotation.tasks.services.get_new_id") as mock_get_new_id:
+        mock_get_new_id.side_effect = (
+            lambda old_id, _: old_id if old_id in [2, 3] else None
+        )
+        result_links, result_children = services.get_links_and_children(
+            common_objs_ids, all_tasks_objs, id_mapping
+        )
+
+        assert result_links == expected_links
+        assert result_children == expected_children
+
+
+@pytest.mark.parametrize(
+    ("items", "expected"),
+    (
+        (
+            [
+                [{"a": 1, "b": 2, "c": 3}, {"d": 4}, {"e": 5}],
+                [{"d": 4}, {"f": 6}],
+                [{"d": 4}, {"g": 7}],
+            ],
+            [{"d": 4}],
+        ),
+        (
+            [
+                [{"a": 1}],
+                [],
+                [{"b": 2}],
+            ],
+            [],
+        ),
+        (
+            [
+                [{"a": 1}, {"b": 2}],
+            ],
+            [{"a": 1}, {"b": 2}],
+        ),
+        ([], []),
+    ),
+)
+def test_get_common_values(items: List[List[Any]], expected: List[Any]):
+    assert services.get_common_values(items) == expected
+
+
+def test_change_ids():
+    all_tasks_objs = {
+        "task1": {"objects": [{"id": 1}, {"id": 2}]},
+        "task2": {"objects": [{"id": 3}, {"id": 4}]},
+    }
+
+    doc_objs = {
+        "doc1": {
+            "objects": [
+                {"id": 1, "links": [2], "children": [3]},
+                {"id": 2},
+                {"id": 3},
+            ]
+        }
+    }
+
+    id_mapping = {(1,): 101, (2,): 102, (3,): 103}
+    mock_get_new_id = MagicMock(
+        side_effect=lambda old_id, _: id_mapping.get((old_id,), old_id)
+    )
+    mock_get_common_ids = MagicMock(
+        side_effect=lambda old_id, _: [id_mapping.get((old_id,), old_id)]
+    )
+    mock_get_links_and_children = MagicMock(
+        side_effect=lambda ids, _, __: (
+            [id_mapping.get((i,), i) for i in ids],
+            [],
+        )
+    )
```

Comment:
seems too complex, could you explain what is happening here?

Reply:
In the test, I used lambda functions in mocks to simulate dynamic ID transformations and relationships. The lambda functions mapped old IDs to new ones, retrieved common IDs, and processed links and children according to the predefined id_mapping. This approach allowed me to test how change_ids handles different ID updates and relationships without requiring complex setup or additional test data.